### PR TITLE
[Merged by Bors] - sql: PoETs

### PR DESCRIPTION
**Pull request by nkryuchkov**
## Motivation

Part of https://github.com/spacemeshos/go-spacemesh/issues/2918

## Changes

- move PoETs storage to SQLite

## Test Plan

UT, ST

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by nkryuchkov** [comment]
bors try

File: sql/migrations/0001_initial.sql [hunk 1]
```diff
@@ -88,3 +88,18 @@ CREATE TABLE atx_top
...
+    poet BLOB
+) WITHOUT ROWID;
+
+CREATE TABLE poet_subscriptions
+(
+    service_id VARCHAR,
+    round_id   VARCHAR,
+    ref        VARCHAR,
+    PRIMARY KEY (service_id, round_id)
+);
```
Comment (dshulyak) [inline sql/migrations/0001_initial.sql:98]: please merge these two tables. to keep things simple ref can remain primary key, and round_id, service_id can be an index if they are queried somewhere.

**Comment by nkryuchkov** [comment]
bors try

File: activation/poetdb.go [hunk 2]
```diff
@@ -83,31 +86,45 @@ func (db *PoetDb) storeProof(proofMessage *types.PoetProofMessage) error {
...
+	if err != nil {
+		return err
 	}
-	key := makeKey(proofMessage.PoetServiceID, proofMessage.RoundID)
-	if err := batch.Put(key[:], ref); err != nil {
-		return fmt.Errorf("failed to store poet proof index entry for poetId %x round %s: %v",
+
+	defer dbtx.Release()
+
+	if err := poets.Add(dbtx, ref, messageBytes, proofMessage.PoetServiceID, proofMessage.RoundID); err != nil {
```
Comment (countvonzero) [inline activation/poetdb.go]: why does this need a dbtx if there is only one stmt to execute?

Comment (nkryuchkov) [reply activation/poetdb.go]: Forgot to remove it after resolving a CR request. Thanks. Removed

File: sql/poets/poets.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
+package poets
+
+import (
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+// GetPoET gets a PoET for a given ref.
+func GetPoET(db sql.Executor, ref []byte) (poet []byte, err error) {
```
Comment (dshulyak) [inline sql/poets/poets.go]: maybe just Get instead of GetPoeT and GetBlob ?

File: activation/poetdb.go [hunk 4]
```diff
@@ -216,3 +213,23 @@ func validatePoet(membershipRoot []byte, merkleProof shared.MerkleProof, leafCou
...
+func (db *PoetDb) PoETs() database.Getter {
+	return newPoETFetcherDB(db)
+}
+
+func newPoETFetcherDB(db *PoetDb) *PoETFetcher {
+	return &PoETFetcher{DB: db}
+}
+
+// PoETFetcher is an adapter of SQLite implementation to legacy LevelDB interfaces.
+// TODO(nkryuchkov): Remove when transition to SQLite is finished.
```
Comment (dshulyak) [inline activation/poetdb.go]: it is not about legacy leveldb, this is an interface for the fetcher.
we probably won't get rid of it in near future

File: sql/migrations/0001_initial.sql [hunk 5]
```diff
@@ -88,3 +88,13 @@ CREATE TABLE atx_top
...
     layer  INT NOT NULL
 );
+
+CREATE TABLE poets
+(
+    ref        VARCHAR PRIMARY KEY,
+    poet       BLOB,
+    service_id VARCHAR,
+    round_id   VARCHAR
+) WITHOUT ROWID;
```
Comment (dshulyak) [inline sql/migrations/0001_initial.sql]: WITHOUT ROWID is not good with big blobs, it reduces tree fanout that makes traversal slower
checkout [part 4](https://www.sqlite.org/withoutrowid.html)

**Review by dshulyak — APPROVED** [review]

File: sql/poets/poets.go (hunk 3 above)
Comment (nkryuchkov) [reply sql/poets/poets.go]: I don't mind

File: activation/poetdb.go (hunk 4 above)
Comment (nkryuchkov) [reply activation/poetdb.go]: I reviewed the fetcher code — yes, you're right. I'll remove the same comment in ATXs.

File: sql/migrations/0001_initial.sql (hunk 5 above)
Comment (nkryuchkov) [reply sql/migrations/0001_initial.sql]: thanks for the link

**Comment by nkryuchkov** [comment]
bors merge

**Pull request closed without merging**